### ISSUE TITLE
v5 canary-sheet: invalid hook call during SSR

### ISSUE THREAD

**Issue by wmertens** [styled-components/styled-components#2608]
## System:
 - OS: macOS 10.14.5
 - CPU: (4) x64 Intel(R) Core(TM) i5-7360U CPU @ 2.30GHz
 - Memory: 41.20 MB / 8.00 GB
 - Shell: 3.2.57 - /bin/bash
## Binaries:
 - Node: 10.15.3 - ~/.nix-profile/bin/node
 - Yarn: 1.16.0 - ~/bin/node_modules/.bin/yarn
 - npm: 6.9.0 - ~/bin/node_modules/.bin/npm
 - Watchman: 4.9.0 - /Users/wmertens/.nix-profile/bin/watchman
## npmPackages:
 - babel-plugin-styled-components: ^1.10.0 => 1.10.0
 - styled-components: ^5.0.0-5.canary-sheet => 5.0.0-5.canary-sheet

## Reproduction

No repro (yet), I'm in a pretty complex project that's hard to minimize. Let's start with theoretical causes?

I upgraded from v4 to the v5 canary, and I got this error:

```
Invariant Violation: Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
1. You might have mismatching versions of React and the renderer (such as React DOM)
2. You might be breaking the Rules of Hooks
3. You might have more than one copy of React in the same app
See https://fb.me/react-invalid-hook-call for tips about how to debug and fix this problem.
    at invariant (myproject/node_modules/.registry.npmjs.org/react/16.8.6/node_modules/react/cjs/react.development.js:88:15)
    at resolveDispatcher (myproject/node_modules/.registry.npmjs.org/react/16.8.6/node_modules/react/cjs/react.development.js:1436:28)
    at Object.useMemo (myproject/node_modules/.registry.npmjs.org/react/16.8.6/node_modules/react/cjs/react.development.js:1491:20)
    at useStyleSheetProvider (myproject/node_modules/.registry.npmjs.org/styled-components/5.0.0-5.canary-sheet/node_modules/styled-components/dist/styled-components.cjs.js:1408:16)
    at StyleSheetManager (myproject/node_modules/.registry.npmjs.org/styled-components/5.0.0-5.canary-sheet/node_modules/styled-components/dist/styled-components.cjs.js:1420:15)
    at processChild (webpack-internal:///./node_modules/.registry.npmjs.org/react-dom/16.8.6/node_modules/react-dom/cjs/react-dom-server.node.development.js:2888:14)
    at resolve (webpack-internal:///./node_modules/.registry.npmjs.org/react-dom/16.8.6/node_modules/react-dom/cjs/react-dom-server.node.development.js:2812:5)
    at ReactDOMServerRenderer.render (webpack-internal:///./node_modules/.registry.npmjs.org/react-dom/16.8.6/node_modules/react-dom/cjs/react-dom-server.node.development.js:3202:22)
    at ReactDOMServerRenderer.read (webpack-internal:///./node_modules/.registry.npmjs.org/react-dom/16.8.6/node_modules/react-dom/cjs/react-dom-server.node.development.js:3161:29)
    at renderToStaticMarkup (webpack-internal:///./node_modules/.registry.npmjs.org/react-dom/16.8.6/node_modules/react-dom/cjs/react-dom-server.node.development.js:3661:27)
    at process (webpack-internal:///./node_modules/.registry.npmjs.org/react-apollo/2.5.5/node_modules/react-apollo/react-apollo.esm.js:1064:20)
    at process._tickCallback (internal/process/next_tick.js:68:7) name: 'Invariant Violation', framesToPop: 1
```

## Steps to reproduce

I add S-C for SSR as follows:

```js
[...]
	const sheet = new ServerStyleSheet()
	renderContext.app = (
		<StyleSheetManager sheet={sheet.instance}>
			{renderContext.app}
		</StyleSheetManager>
	)
	try {
		await hooks.prerender.promise(renderContext)
		hooks.afterPrerender.call(renderContext)
		if (renderContext.redirect) return {redirect: renderContext.redirect}

		let content = renderToString(renderContext.app)

		hooks.afterRender.call(renderContext)
		if (renderContext.result) return renderContext.result

		const styles = sheet.getStyleElement()
[...]
```

So basically wrap the app with StyleSheetManager and the sheet instance, and then get the styles with getStyleElement. It renders twice, once in the prerender hooks via Apollo's `getDataFromTree`, and then again on the `renderToString`.

Thoughts? The client side works fine, it's so nice to have a shallower vDOM :)

**Comment by probablyup** [comment]
That cut is actually pretty old now with all the changes we have made in
the past 2 weeks. Can I do another pre alpha release for you to try out?

**Comment by wmertens** [comment]
Yes please :)

**Comment by probablyup** [comment]
@wmertens give `styled-components@alpha` a try :)

**Comment by probablyup** [comment]
@wmertens if it continues to exist, a reproduction I could pop into a jest test would be awesome to TDD a fix :)

**Comment by probablyup** [comment]
@wmertens are you seeing in this in the beta, or can I close?

**Comment by wmertens** [comment]
@probablyup 	I'm sorry, I missed all these notifications :( still happening in beta.5

**Comment by wmertens** [comment]
I think I know why: react-apollo v2.5.8 getDataFromTree seems to be causing it. So maybe this isn't an issue we can resolve. Keeping it open for just a bit longer while I investigate.

**Comment by wmertens** [comment]
OK I'm pretty sure it's an issue with react-apollo https://github.com/apollographql/react-apollo/issues/3071

**Comment by probablyup** [comment]
Fun! Thanks for following up 😄

**Comment by kylemh** [comment]
I'm seeing this with a Next.js app (no `react-apollo`) with `v4@latest` of `styled-components`

**Comment by probablyup** [comment]
@kylemh we don't use hooks in v4 so that's basically impossible

**Comment by wmertens** [comment]
@kylemh 	maybe you're conditionally calling hooks somewhere?

**Comment by kylemh** [comment]
@probablyup oh to be clear wasn't blaming `styled-components` source code! I misread this issue. I presumed it was about a compatibility issue between hooks-using components that are also `styled-components`.

**Comment by wmertens** [comment]
I finally tracked down the problem: in the server build, react-dom was internalized, but styled-components was external, so they actually used two different instances of the same `react` :man_facepalming:

**Comment by JeffWeim** [comment]
@wmertens Would you be able to elaborate any more on what you did to resolve your issue?

**Comment by wmertens** [comment]
@JeffWeim I made sure that react-dom and styled-components used the same `react` instance

**Comment by kitten** [comment]
Can you try this with react-ssr-prepass instead please? Instructions are in the readme: https://github.com/FormidableLabs/react-ssr-prepass

If the same bug occurs there it’ll be easier to read the stack trace

**Comment by Dip686** [comment]
@wmertens I am facing similar issue, 
setting `'react': path.resolve('./node_modules/react')` in webpack is resolving the issue by forcing to refer the same react package. Have you taken a similar approach or tried something else?

**Comment by wmertens** [comment]
@Dip686 I found where the duplication was and eliminated it. I think it was due to webpack bundling react in the server script but not for all dependencies.

**Comment by Dip686** [comment]
Sharing my use case, someone may find it useful,
I was creating my own Package(uses react and styled-components internally) when I am trying to use it in our orgs existing app it was providing the same error. My use-case is Org specific and will be published for org only. Also, the build is only transpiled code, not fully bundled as the existing App is bundling. Found that the App had a different version of React, prop-types, etc. My lib's package.json earlier generated with react, styled-components,prop-types as a dependency. This was creating a problem. Hence I decided to change it to peerDependency.
